accepted/2020/diagnostics/activity-improvements-2.md

```diff
@@ -0,0 +1,156 @@
+
+ActivityTagsCollection is needed for the following reasons:
+
+- We need to avoid having the consumers of our APIs to use any other types (e.g. Dictionary<>, List<>...etc.) which can cause issues like not keeping the items in order or not conforming to the OT specs which suggest specific behaviors (e.g. when adding null value, should remove the entry from the list) 
+- Need to support the nested attributes. ActivityTagsCollection will make it easy to do so.
+
+```c#
+    public class ActivityTagsCollection : IReadOnlyCollection<KeyValuePair<string, object>>
+    {
+        public ActivityTagsCollection() { }
+        public ActivityTagsCollection(IEnumerable<KeyValuePair<string, object>>) { }
+
+        public void Add(string key, int value) { }
+        public void Add(string key, double value) { }
+        public void Add(string key, bool value) { }
+        public void Add(string key, string value) { }
+        public void Add(string key, int [] value) { }
+        public void Add(string key, double [] value) { }
+        public void Add(string key, bool [] value) { }
+        public void Add(string key, string [] value) { }
+        public void Add(string key, ActivityTagsCollection value) { }
+
+        // interfaces implementation
+        public int Count { get; }
+        public IEnumerator<KeyValuePair<string, object>> GetEnumerator() { ... }
+        IEnumerator IEnumerable.GetEnumerator() { ... }
+    }
+```
+
+## Activity.Tags
+
+As mentioned earlier, Activity has Tags property which returns the list of the activity tags. Tags are a list of `KeyValuePair<string, string>`. Tags are proven to be not enough for OT scenarios as we need to support more value types (numeric, bool, numeric array, and nested tags). Activity today has the method `public Activity AddTag(string key, string? value)`. The proposal here is to add more `AddTag` overload accepting more types and add a new getter property return the list of all typed tags. 
+
+```C#
+    public partial class Activity : IDisposable
+    {
+        public void AddTag(string key, int value) { }
+        public void AddTag(string key, double value) { }
+        public void AddTag(string key, bool value) { }
+        public void AddTag(string key, int [] value) { }
+        public void AddTag(string key, double [] value) { }
+        public void AddTag(string key, bool [] value) { } // how interesting this one?
+        public void AddTag(string key, string [] value) { }
+        public void AddTag(string key, ActivityTagsCollection value) { }
+
+        // We cannot change Tags property so we are introducing a new property with a new name
+        public ActivityTagsCollection TypedTags { get; }
```

Comment:
Clarification question - if we enumerate through the `TypedTags`, do we get the string tag added via `AddTag(string, string)` or we have to do another enumeration through `Tags`?

Reply:
TypedTags should return all tags including the tags added with `AddTag(string, string)`